[Build] Add the new fully-static Linux SDK.

al45tair opened the pull request:
Declare a new `LINUX_STATIC` SDK and configure it.
    
Add options to set the build architectures for the `LINUX` and `LINUX_STATIC` SDKs, similar to what we have for Darwin, because we'll be cross-compiling.

Also add an option to point the build system at the sources for the musl C library, which we're using for `LINUX_STATIC`.

rdar://123503470

al45tair committed:
[Build] Add the new fully-static Linux SDK.

Declare a new `LINUX_STATIC` SDK and configure it.

Add options to set the build architectures for the `LINUX` and
`LINUX_STATIC` SDKs, similar to what we have for Darwin, because
we'll be cross-compiling.

Also add an option to point the build system at the sources for
the musl C library, which we're using for `LINUX_STATIC`.

rdar://123503470

al45tair commented:
This sits on top of #71838.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

edymtt commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
I'm realizing this is the trickiest file to review, since it has a sizable amount of logic (and a lot of duplication) and I don't have all the use cases in mind -- so I expect my comments to be pointless or to be refuted by failures in PR testing.

edymtt reviewed (commented):
The overall approach looks good, given the moderate size I did a first pass with my initial impressions -- loosely wondering if there is a way to break this PR in smaller chunks (e.g. land first the plumbing of the build-script args, and leave last the more substantial changes to the CMake logic that may require more discussion).

al45tair commented on CMakeLists.txt:
```diff
@@ -1059,9 +1074,25 @@ if("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "LINUX")
   # Should we build the standard library for the host?
```
Fair point, though that isn't new.

al45tair commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -1336,10 +1344,14 @@ function(add_swift_target_library_single target name)
-        target_static_depends)
+    set(target_static_depends)
+    foreach(dep ${SWIFTLIB_SINGLE_LINK_LIBRARIES})
+      if (NOT "${dep}" MATCHES "^(icucore|dispatch|BlocksRuntime)($|-.*)$")
```
I see what you're saying, however this code isn't gated on `NOT SWIFTLIB_OBJECT_LIBRARY` or `SWIFTLIB_IS_STDLIB`. Given the timeline I'm not keen on removing it at this point — there's the potential for unexpected fallout, as you say — but we should probably consider it at some point in the future (assuming @etcwilde doesn't beat us to it with an entirely new set of CMake scripts).

al45tair commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -1885,6 +1899,17 @@ function(add_swift_target_library name)
   endif()
   list_replace(SWIFTLIB_TARGET_SDKS ALL_APPLE_PLATFORMS "${SWIFT_DARWIN_PLATFORMS}")
 
+  # Support adding a "NOT" on the front to mean all SDKs except the following
```
It's used in a later PR in this set. I'd have to dig about to figure out which one, but we do need it and this seemed like a not unreasonable place to add it.

al45tair commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -2341,23 +2378,19 @@ function(add_swift_target_library name)
-          if (NOT "${DEP}" STREQUAL "icucore" AND
-              NOT "${DEP}" STREQUAL "dispatch" AND
-              NOT "${DEP}" STREQUAL "BlocksRuntime")
+        foreach(dep ${SWIFTLIB_LINK_LIBRARIES})
+          if (NOT "${dep}" MATCHES "^(icucore|dispatch|BlocksRuntime)($|-.*)$")
```
Personally I find the single line version easier to follow (and it only does one match instead of three), but not changing it does result in a smaller diff, so let's leave it.

al45tair commented on utils/build-script:
```diff
@@ -297,6 +297,12 @@ def default_stdlib_deployment_targets(args):
             targets.extend(StdlibDeploymentTarget.AppleWatch.targets)
         return targets
+    elif host_target.platform.name == 'linux':
+        targets = [host_target]
```
Yeah, that does make things cleaner.

al45tair commented on utils/build-script-impl:
```diff
@@ -1392,6 +1401,14 @@ function swift_c_flags() {
     if [[ $(is_cmake_release_build_type "${SWIFT_BUILD_TYPE}") ]] ; then
         echo -n " -fno-stack-protector"
     fi
+
```
I like the idea, but I think it's a riskier change so I've filed a separate radar for it: rdar://127421155

al45tair commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -1325,6 +1335,28 @@ def create_argument_parser():
+    # -------------------------------------------------------------------------
+    in_group('Build settings for Linux')
+
+    option('--linux-arch', store,
```
Have updated this everywhere.

al45tair commented on utils/swift_build_support/swift_build_support/products/llvm.py:
```diff
@@ -351,6 +352,36 @@ def build(self, host_target):
             llvm_cmake_options.define('LLVM_INCLUDE_TESTS', 'NO')
             llvm_cmake_options.define('CLANG_INCLUDE_TESTS', 'NO')
 
+        build_root = os.path.dirname(self.build_dir)
```
We will want them in the generated toolchains.

al45tair replied:
I'm slightly torn about this; I had intended to get rid of the `.cfg` files by patching the compiler drivers, particularly now I've changed the triple to a special one for the static SDK (`<arch>-swift-linux-musl` instead of `<arch>-unknown-linux-musl`). I think though that we probably do want to add them in the install phase for now; we can always remove them again later if it's a problem.

al45tair committed:
[Build] Updated following initial comments from Eric.

Change the `--linux[-static]-arch` option to `--linux[-static]-archs`, on the
basis that it supports multiple values.

Other tidying.

rdar://123503470

al45tair committed:
[Build] Fix a conflict resolution mistake.

Apparently I accidentally removed a `]` while resolving a conflict in
`build_script_invocation.py.`

rdar://123503470

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Build] Fix Python linting issues.

The Python linter was complaining about import orders.

rdar://123503470

al45tair committed:
[Build] Add options to the Python build script tests.

Apparently I'd failed to add a couple of lines here.

rdar://123503470

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented on utils/build-script-impl:
```diff
@@ -130,6 +135,8 @@ KNOWN_SETTINGS=(
     skip-build                                    ""                "set to configure as usual while skipping the build step"
     skip-build-android                            ""                "set to skip building Swift stdlibs for Android"
+    skip-build-linux                              ""                "set to skip building Swift stdlibs for Linux"
+    skip-build-linux-static                       ""                "set to skip building Swift stdlibs for statically linked Linux"
```
There is precedent for the inversion here; I was just copying what the Android option does (that's `--skip-build-android` at `build-script-impl` level but `--build-android` for `build-script`). But yes, perhaps it does make more sense not to invert it and to just regard the Android thing as a mistake.

al45tair committed:
[Build] Further static Linux SDK support changes.

Fix some indentation issues.

Change `build-script-impl` to make `build-linux-static` a positive argument.

Fix documentation for `--linux-archs` and `--linux-static-archs` (the options
are comma separated for `build-script`, but semicolon separated for
`build-script-impl`).

Set the default for `linux-static-archs` to `x86_64, aarch64` so that we
install the expected content in the toolchain.

Add missing default for `test_linux_static`.

Make sure to pass down `--skip-build-linux` and `--build-linux-static`.

Factor out config file generation and call it from the install step in `llvm.py`
as well as from the build step.

rdar://123503470

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Build] A couple more Python fixes.

Remove unused import of `importlib.resources`.

Set `test_linux_static` to `False` if tests are disabled.

rdar://123503470

MaxDesiatov commented:
@swift-ci test webassembly

al45tair commented:
Failures are:
```
Failed Tests (1):
  Swift(linux-x86_64) :: TypeRoundTrip/round-trip.swift
```
because
```
/usr/bin/ld.gold: error: cannot find -lswiftRemoteInspection
/usr/bin/ld.gold: internal error in read_header_prolog, at ../../gold/dwarf_reader.cc:1678
clang: error: linker command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)
```
(similar error on macOS).

Not sure this is to do with this PR; looking into it now.

al45tair commented:
`build-script -r; utils/run-test --build-dir ... test/TypeRoundTrip/round-trip.swift` works locally on macOS, with an initially empty build directory.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Hmmm. It's failing because it can't find `swiftRemoteInspection`, which suggests we're missing a dependency somewhere.

al45tair committed:
[Build][Test] Fix TypeRoundTrip/round-trip.swift.

`TypeRoundTrip/round-trip.swift` tries to link with `libswiftRemoteInspection`,
but that library isn't in the default link path, so we need to add a `-L`
option to the command line in the test.

Also, we should check that `SWIFT_BUILD_REMOTE_MIRROR` is enabled and
disable the test if not, because otherwise `libswiftRemoteInspection` won't
have been built.

rdar://123503470

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Build] Only add -lfts for linux-static-*.

Looks like I'd unconditionally added -lfts to the Foundation build flags.

rdar://123503470

al45tair commented:
@swift-ci Please smoke test Linux platform

al45tair commented:
(The previous failure was because the integration tests tried to treat the `.cfg` files as ELF binaries. That's wrong, so I fixed the integration tests to ignore them.)
